### PR TITLE
Consistent aspect ratio of book covers in mosaic menu

### PR DESCRIPTION
**Pull request by tymmej**
#11835

This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11838)

### CODE REVIEW

File: plugins/coverbrowser.koplugin/main.lua [hunk 1]
```diff
@@ -368,6 +368,42 @@ function CoverBrowser:addToMainMenu(menu_items)
                     }
                 }
             },
+            {
+                text = _("Consistent aspect ratio of book covers"),
+                checked_func = function()
+                    return BookInfoManager:getSetting("consistent_aspect_ratio_covers")
```
Comment (poire-z) [inline plugins/coverbrowser.koplugin/main.lua:374]: I think it needs some other wording. "Consistent" sounds like a good thing to me - and seeing that wording I would probably check it, while here it means that the user will get a consistently screwed up aspect ratio :)

File: plugins/coverbrowser.koplugin/mosaicmenu.lua [hunk 2]
```diff
@@ -585,12 +585,24 @@ function MosaicMenuItem:update()
                 cover_bb_used = true
                 -- Let ImageWidget do the scaling and give us a bb that fit
                 local _, _, scale_factor = BookInfoManager.getCachedCoverSize(bookinfo.cover_w, bookinfo.cover_h, max_img_w, max_img_h)
+                local consistent_aspect_ratio_covers = BookInfoManager:getSetting("consistent_aspect_ratio_covers") and true or false
+
+                local scale_factor_img = consistent_aspect_ratio_covers and nil or scale_factor
+                if consistent_aspect_ratio_covers then scale_factor_img = nil end
```
Comment (poire-z) [inline plugins/coverbrowser.koplugin/mosaicmenu.lua:591]: I think you got there by _trial and error and add kludge to fix the error_ :)
I guess what you want is just:
`local scale_factor_img = not consistent_aspect_ratio_covers and scale_factor or nil`

File: plugins/coverbrowser.koplugin/mosaicmenu.lua [hunk 3]
```diff
@@ -585,12 +585,24 @@ function MosaicMenuItem:update()
...
+                local stretch = BookInfoManager:getSetting("aspect_ratio_covers_max_stretch") or 8
+                stretch = consistent_aspect_ratio_covers and stretch or nil
+
                 local image= ImageWidget:new{
                     image = bookinfo.cover_bb,
-                    scale_factor = scale_factor,
+                    scale_factor = scale_factor_img,
+                    stretch_limit_percentage = stretch,
+                    width = consistent_aspect_ratio_covers and max_img_w or nil,
+                    height = consistent_aspect_ratio_covers and max_img_h or nil,
```
Comment (poire-z) [inline plugins/coverbrowser.koplugin/mosaicmenu.lua:601]: There is something a bit more serious, writing about it in the main discussion.

File: plugins/coverbrowser.koplugin/main.lua (hunk 1 above)
Comment (Frenzie) [reply plugins/coverbrowser.koplugin/main.lua:374]: Forced?

**Comment by poire-z** [comment]
(And this will conflict with #11792 - so if any of these ends up being accepted, you may already think about how it should integrate in the future menu layout.)

**Comment by NiLuJe** [comment]
Yeah, this will absolutely murder filemanager performance.

**Comment by NiLuJe** [comment]
I still have the notes I took during the aforementioned refactor from @hius07, so I *may* indeed get to do *something* about it at one point (my original intent being able to keep *multiple* thumbnails, one for each view mode), which would potentially make this slightly less awkward.

I'm still *extremely* "meh" about the whole concept of this PR, though ;).

File: plugins/coverbrowser.koplugin/main.lua (hunk 1 above)
Comment (NiLuJe) [reply plugins/coverbrowser.koplugin/main.lua:374]: I'd possibly wnt to stack *both* force + consistent, *and* add an `help_text` to confirm that this does indded mean "strech things to make everything look like ass" ;p (/s about the opinionated wording, but I do mean the first bit, e.g., an help_text that mpentions stretching).

Comment (NiLuJe) [reply plugins/coverbrowser.koplugin/main.lua:374]: And of course that it'll tank browsing performance, see main discussion.

**Comment by hius07** [comment]
To avoid the slowdown we must cache the streched thumbnails
https://github.com/koreader/koreader/blob/d801af6d418e3b4f1ad090047e4370d9f3d198ad/plugins/coverbrowser.koplugin/bookinfomanager.lua#L518-L522
`BookInfoManager.getCachedCoverSize()` should consider the new setting.

**Comment by NiLuJe** [comment]
That's assuming the stretch factor isn't variable; but, yeah, that'd be less terrible from a UX standpoint, potentially messier from a code quality standpoint ;).

**Comment by offset-torque** [comment]
Can't this be a user script? No extras like caching, tagging, performance concerns etc. No need for a menu even. If they can install a user script they can change the values directly in there with the help of some explanatory comments. 

Users bothered with the current AR can try it. And if they think performance loss is worth it, they can keep using it. Otherwise they can disable and go back to original AR. And **tymmej** can update the script if necessary as long as he uses KOReader.

**Comment by hius07** [comment]
> Can't this be a user patch?

I'm afraid no. I'm going to look closer at it after merging (or rejecting) https://github.com/koreader/koreader/pull/11792.

**Comment by poire-z** [comment]
> Can't this be a user script?

It was hard making one :) but yes, it could.

I would not really care about the performance and quality: we all, including the requesters of this, know and are ok that it is bad quality and looks like ass (as it is already distorted), so less ass or more ass... I'd rather keep it bad and slow, than have this leak and need tweaks in many other things like cached bb, fake covers...
And it would need more tweaks, ie. the Fake/text covers (that one can get with books without covers) are currently not sized accordingly by this PR.
I still feel it looks bad: the OP in his screenshots in #11835 has only colorfull adequately sized covers, where it looks fine. But it you have a tall cover exceeding the stretch %, and if it is a light color, and does not have itself borders, you end up with strange looking stuff. At least, our own framecontainer borders add a border to covers who don't have one.
It also looks fine only when your grid NxM, and your device screen aspect ratio, are chosen to in a way that matches a regular book aspect ratio (which ok, we probably all have it so), so it still needs a bit of fine tuning by the user, and won't adjust well when grid NxM are changed (which ok, we probably never change it when satistifed).

So, anyway, here's a user patch - so at least people can try it, see how it feels and what's missing, so that if they end up really satisfied with it we can think again about it. But let's wait a bit before getting this leaking and needing more tweaks in the rest of our stuff.
[2-CoverBrowserMosaicStretched.lua.txt](https://github.com/koreader/koreader/files/15368567/2-CoverBrowserMosaicStretched.lua.txt)
```lua
local ImageWidget = require("ui/widget/imagewidget")
local PluginLoader = require("pluginloader")
local Size = require("ui/size")

-- Set your max bearable stretch level
local STRETCH_LIMIT_PERCENTAGE = 20

-- On each new Reader/FileManager, plugins are dofile()d, and then
-- instantiated thru createPluginInstance: catch there ere.
local patch_plugin_func -- defined below
local orig_PluginLoader_createPluginInstance = PluginLoader.createPluginInstance
PluginLoader.createPluginInstance = function(self, plugin, attr)
    local ok, plugin_or_err = orig_PluginLoader_createPluginInstance(self, plugin, attr)
    if ok and plugin.name == "coverbrowser" then
        patch_plugin_func(plugin)
    end
    return ok, plugin_or_err
end

-- A bit tedious to reach and tweak MosaicMenuItem's ImageWidget usage...
patch_plugin_func = function(plugin)
    -- In our case, nothing much to do with the plugin object we get passed

    local MosaicMenu = require("mosaicmenu")

    -- We need to find the module's local ImageWidget (= require("ui/widget/imagewidget")),
    -- We need to jump thru 2 objects' methods to reach it from their upvalues...
    local MosaicMenuItem
    local n = 1
    while true do
        local name, value = debug.getupvalue(MosaicMenu._updateItemsBuildUI, n)
        if not name then break end
        if name == "MosaicMenuItem" then
            MosaicMenuItem = value
            break
        end
        n = n + 1
    end
    if not MosaicMenuItem then return end

    local ImageWidget
    local n = 1
    while true do
        local name, value = debug.getupvalue(MosaicMenuItem.update, n)
        if not name then break end
        if name == "ImageWidget" then
            ImageWidget = value
            break
        end
        n = n + 1
    end
    if not ImageWidget then return end
    local setupvalue_n = n -- we will replace it with a small subclass

    -- We can't access max_img_w/h, which are defined in MosaicMenuItem.update
    -- for each instance. We need to intercept each instantiation via :init()
    -- and compute them the same way they are computed in MosaicMenuItem.update
    local max_img_w, max_img_h
    local border_size = Size.border.thin -- defined in MosaicMenuItem:update()
    local underline_h = 1 -- defined in MosaicMenuItem:init()
    local orig_MosaicMenuItem_init = MosaicMenuItem.init
    MosaicMenuItem.init = function(self)
        -- Witnessed a crash where self.width was nil, not sure how,
        -- so better check and do less well, but better than crashing.
        if self.width and self.height then
            -- We compute and set our local upvalues, that will be used by StretchingImageWidget
            max_img_w = self.width - 2*border_size
            max_img_h = self.height - 2*border_size - underline_h
        end
        orig_MosaicMenuItem_init(self)
    end

    -- Small subclass of ImageWidget to force the setting we could just have inserted
    -- in mosaicmenu, as in top post of https://github.com/koreader/koreader/issues/11835
    local StretchingImageWidget = ImageWidget:extend{}
    -- (ImageWidget has no :init(), so no tedious gymnastic needed)
    StretchingImageWidget.init = function(self)
        if not max_img_w and not max_img_h then
            -- As above, do nothing if we were not able to compute them
            return
        end
        self.scale_factor = nil -- reset the one set
        self.width = max_img_w  -- use the one computed above
        self.height = max_img_h
        self.stretch_limit_percentage = STRETCH_LIMIT_PERCENTAGE
    end

    -- Have mosaicmenu.lua's local ImageWidget be our StretchingImageWidget
    debug.setupvalue(MosaicMenuItem.update, setupvalue_n, StretchingImageWidget)
end
```

**Comment by hius07** [comment]
That's cool.

**Comment by noxhirsch** [comment]
@poire-z Thank you for your hard work on that user patch! Based on your way of patching plugins, I created a user patch for BookInfoManager.getCachedCoverSize (what I did was more trial and error than knowing what I'm doing here - but somehow I got it working). This should in theory work around the performance issues. However, it is of course not possible to adjust text covers that way. Maybe it can be combined with your patch and extended.

```lua
local PluginLoader = require("pluginloader")

local STRETCH_LIMIT_PERCENTAGE = 20

local patch_plugin_func -- defined below
local orig_PluginLoader_createPluginInstance = PluginLoader.createPluginInstance
PluginLoader.createPluginInstance = function(self, plugin, attr)
    local ok, plugin_or_err = orig_PluginLoader_createPluginInstance(self, plugin, attr)
    if ok and plugin.name == "coverbrowser" then
        patch_plugin_func(plugin)
    end
    return ok, plugin_or_err
end

patch_plugin_func = function(plugin)
    local BookInfoManager = require("bookinfomanager")
    
    BookInfoManager.getCachedCoverSize = function(img_w, img_h, max_img_w, max_img_h)
        local scale_factor = max_img_w / img_w
        local max_img_w = max_img_w
        local max_img_h = max_img_h

        local max_img_ratio = max_img_w / max_img_h
        local image_ratio = img_w / img_h
        local ratio_divergence_percent = math.abs(100 - image_ratio / max_img_ratio * 100)
        if ratio_divergence_percent > STRETCH_LIMIT_PERCENTAGE then
            local width = math.floor(max_img_h * img_w / img_h + 0.5)
            if max_img_w >= width then
                max_img_w = width
                scale_factor = max_img_w / img_w
            else
                max_img_h = math.floor(max_img_w * img_h / img_w + 0.5)
                scale_factor = max_img_h / img_h
            end
        end
        return max_img_w, max_img_h, scale_factor
    end
end
```

**Comment by poire-z** [comment]
I'll probably add to our userpatch module little helper functions available for use in user patches, so we don't have to duplicate my 10+-lines tricks; for example, my above patch first 45 lines could be reduced to:
```lua
local userpatch = require("userpatch")
-- A bit tedious to reach and tweak MosaicMenuItem's ImageWidget usage...
userpatch.registerPatchPluginFunc("coverbrowser", function(plugin)
    -- In our case, nothing much to do with the plugin object we get passed

    local MosaicMenu = require("mosaicmenu")
    -- We need to find the module's local ImageWidget (= require("ui/widget/imagewidget")),
    -- We need to jump thru 2 objects' methods to reach it from their upvalues...
    local MosaicMenuItem = userpatch.getUpValue(MosaicMenu._updateItemsBuildUI, "MosaicMenuItem")
    if not MosaicMenuItem then return end

    local ImageWidget, upvalue_idx = userpatch.getUpValue(MosaicMenuItem.update, "ImageWidget")
    if not ImageWidget then return end
end)
...

```

**Comment by poire-z** [comment]
Closing for now, until we eventually get more time to think about it and the other stuff that need adjustments.
Two user patches are available above for people who really wants that.

**Pull request closed without merging**